### ISSUE TITLE
FAQs about documents for the team project

### ISSUE THREAD

**Issue by aqjune** [snu-sf-class/swpp202001#12]
1. One person can implement multiple optimizations in a single sprint!
Also, one optimization can be implemented through multiple sprints with your teammates.

2. All modifications other than adding the pass to PassManager is assumed to be implementing a new optimization.
For example, adding a custom optimization (e.g. x + x -> x * 2) that may not exist in InstCombine is a new optimization.
However, adding only one small transformation during one sprint might not be a good idea. Note that TAs will check whether each student is contributing to the team project enough.

3. If you are adding an optimization that was not on the list, please let TAs know.
Then, TAs will check whether it can be implemented with the existing LLVM libraries.
TAs would like to check that students didn’t simply copy & paste the code from LLVM & claim that they implemented something.

**Comment by aqjune** [comment]
Update to the answer to question 2: Making a pass that simply calls library functions (e.g. `llvm::InlineFunction()`) is also assumed to be working on an existing LLVM optimization.

**Comment by aqjune** [comment]
Note that you still need to make your pull request not bigger than 200 lines (as described) and should have one consistent theme.
If you are working on existing optimizations, you can make at most 2 pull requests. Other than this case, you can still make 2, but please try to make only one pull request

**Comment by YPH95** [comment]
Do we need to write a complete IR program or is it enough to focus on only the optimized parts and give a before after kind of look?

**Comment by aqjune** [comment]
It is okay to give a high-level example only.

**Comment by pullup-dip-burpee** [comment]
조교님 안녕하세요, optList 1번부터 7번까지는 자유롭게 구현이 가능하고 8-17번까지는 이미 있는 구현이라 한 sprint 당 한 명만 담당해서 할 수 있는 것인가요? 혹은 8-17번은 한 명이 한 스프린트에 여러 개를 가져와도 되는 것인지 여쭤봅니다.

**Comment by aqjune** [comment]
@pullup-dip-burpee 

1. Yes.

2. Yes, one person can make a pull request that has multiple existing optimizations. Note that it still should not break the condition of a good pull request: one topic and less than 200 lines.

3. load/stores have a cost that is proportional to the travel distance of the head, whereas reset heap only has a fixed cost. If reset heap is appropriately used, the cost of load/store can be reduced.

**Comment by pullup-dip-burpee** [comment]
> Note that you still need to make your pull request not bigger than 200 lines (as described) and should have one consistent theme.
> If you are working on existing optimizations, you can make at most 2 pull requests. Other than this case, you can still make 2, but please try to make only one pull request

조교님 혹시 풀 리퀘스트 200줄, 최대 2회가 cpp코드에만 해당되는 것인지, 아니면 시험용 테스트파일 혹은 다른 markdown 파일도 다 포함한 것인지 여쭤봅니다

**Comment by aqjune** [comment]
It excludes documents and tests (google test, FileCheck).

**Comment by micmic123** [comment]
> [document.pdf]
> C. Progress report
> ...
> i. Test results should include change in cost before and after optimization
> ii. Write down at least 3 IR programs per optimization pass

document파일의 progress report부분을 보면 "optimization 전/후로 test 결과 cost가 달라져야한다"고 되어있습니다. 제가 구현한 최적화가 조교님께서 제공해주신 [test 파일](https://github.com/snu-sf-class/swpp202001-test)에는 해당사항이 없어서 cost 변화가 없는데요, 그러면 제가 따로 만든 test 파일에서 cost 변화가 생기는 것을 확인하고 report하면 될까요?

**Comment by aqjune** [comment]
TAs will not count a pull request that only updates documents / test scripts / copies files from the project skeleton.

**Comment by kwonsw055** [comment]
Pull request에서 200줄 제한은 원본 컴파일러의 버그 픽스 Commit을 cherry picking 한 줄도 포함하는 건가요?

**Comment by aqjune** [comment]
It should have been made as a separate PR... then TAs wouldn't have counted it.

**Comment by kwonsw055** [comment]
실습 시간 자료 중 5.code-review.pdf 슬라이드 18쪽에서
> 200 lines without tests / comments is assumed to be ‘big’

라 명시되어 있는데, 그럼 이번 프로젝트의 PR에서도 주석은 200줄 제한에 포함되지 않는다고 생각하면 될까요?

**Comment by aqjune** [comment]
Comments will be counted as well.
In the practice session I said so, but professor recommended that lines that are self-explanatory don't need comments, so I think the comments should also be counted for this purpose.

**Comment by kangz12345** [comment]
> If you are working on existing optimizations, you can make at most 2 pull requests. Other than this case, you can still make 2, but please try to make only one pull request

매 sprint당 2개의 PR을 할 수 있다는 게 팀원 각자에 해당하는 것인가요? 예를 들어 각자 하나의 서로 다른 최적화를 구현한다면 반드시 팀단위로는 최소 3-4개의 PR이 만들어져야 할 것 같습니다.

**Comment by aqjune** [comment]
Yes, it is per a student.

**Comment by lego0901** [comment]
Progress report에 대해 질문드립니다.
각 최적화에 대한 설명을 적을 때, 세부적인 구현 방법에 대해서도 자세히 적는게 좋을까요?
아니면 표면적으로 이 최적화가 어떤 역할을 하고, 어떤 조건에서만 작동하는지 정도만 명시를 하는게 좋을까요?
얼마나 상세히 적어야할지 감이 잘 안 잡혀서 질문드립니다..

**Comment by aqjune** [comment]
In a progress report, giving a high level description about your implementation is enough.

**Comment by aqjune** [comment]
BTW, please be careful not to include absolute path in your Makefile or anywhere else - this will make building/testing the project on TA's machines fail.

**Comment by tlvkfn525** [comment]
혹시 2개의 연속된 pull request를 한 sprint에서 만들었으면, comment하고 merge하는 과정을 어떻게 처리하는 게 좋을까요? 코딩기간에 comment를 하고 merge를 한 다음에 다음 pull request를 코딩 기간 내에 해야 하는 건지, 아니면 그냥 pull request를 두개 동시에 열어두고 스프린트가 다 끝날 때 한번에 merge를해야 하는지 궁금합니다.

**Comment by aqjune** [comment]
The two PRs created during the implementation period should not be dependent on each other. Merge should be done at the rebuttal period.

**Comment by aqjune** [comment]
If there is a reason that one PR should be dependent on another, ask TA with the detail of the PRs.

**Comment by robinshins** [comment]
혹시 sprint 2 progress report에서 최적화 전의 cost는 아무 최적화도 넣지 않은 상태의 cost를 넣어야할까요 아니면 sprint1 때의 cost를 넣어야 할까요?

**Comment by aqjune** [comment]
Either is fine : you can compare the result to the performance of the previous sprint or the skeleton.

**Comment by StuartSul** [comment]
안녕하세요. 혹시 이제 Sprint 3 도큐먼트 (Spec & Progess) 제출을 제외하면 더 이상 제출할 게 없는 것이 맞나요?

**Comment by aqjune** [comment]
Yes!

**Comment by GGlifer** [comment]
조교님, 혹시 Sprint3 에서 제출할 도큐먼트 중 Requirement & Specification 에는 어떤 내용이 들어가야 하나요?

Describe optimization passes that you are going to be implemented

라고 가이드라인에 적혀져 있는데... 그러면 마지막으로 Wrap Up 때 무엇을 할 것인지 적으면 되나요?

아니면 Sprint 3 때 이러이러한 알고리즘을 구현하고자 했으나 반 정도 밖에 구현하지 못하였다면
그 반 정도의 진도에 맞게 Requirement & Specification 을 수정해야하는 건가요?

**Comment by aqjune** [comment]
> 라고 가이드라인에 적혀져 있는데... 그러면 마지막으로 Wrap Up 때 무엇을 할 것인지 적으면 되나요?

Yes. If there's no plan to modify the implementation, you can just omit it or leave just it as 'there's no plan to change implementation'.

> 아니면 Sprint 3 때 이러이러한 알고리즘을 구현하고자 했으나 반 정도 밖에 구현하지 못하였다면
그 반 정도의 진도에 맞게 Requirement & Specification 을 수정해야하는 건가요?

If there is no plan to do work further (there might be various reasons, such as the performance gain already seemed to be enough), it is enough to submit the require & specification of the previous week, with mentioning in the document that there was no change.
